daemon/mgr/container_checkpoint.go

```diff
@@ -0,0 +1,107 @@
+package mgr
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/alibaba/pouch/apis/types"
+)
+
+// getCheckpointDir gets container checkpoint directory.
+func (mgr *ContainerManager) getCheckpointDir(container, id string, create bool) (string, error) {
+	if id == "" {
+		return "", fmt.Errorf("checkpoint can not be empty")
+	}
+	dir := filepath.Join(mgr.Config.HomeDir, "containerd/state", "io.containerd.runtime.v1.linux", mgr.Config.Namespace, container, "checkpoint", id)
+	stat, err := os.Stat(dir)
+	if create {
+		if err != nil && os.IsNotExist(err) {
+			return dir, os.MkdirAll(dir, 0700)
+		}
+		if err != nil {
+			return "", fmt.Errorf("failed to create checkpoint %s", id)
+		}
+		if !stat.IsDir() {
+			return "", fmt.Errorf("checkpoint %s exist but not directory ", id)
+		}
+		return "", fmt.Errorf("checkpoint %s is already exist", id)
+	}
+
+	if err == nil && stat.IsDir() {
+		return dir, nil
+	}
+	return "", fmt.Errorf("checkpoint %s is not exist", id)
+}
+
+// CreateCheckpoint creates a checkpoint from a running container
+func (mgr *ContainerManager) CreateCheckpoint(ctx context.Context, name string, options *types.CheckpointCreateOptions) error {
+	c, err := mgr.container(name)
+	if err != nil {
+		return err
+	}
+
+	if c.State.Status != types.StatusRunning {
+		return fmt.Errorf("can not checkpoint from a %s container", c.State.Status)
+	}
+
+	if c.Config.Tty {
+		return fmt.Errorf("checkpoint not support on containers with tty")
+	}
+
+	dir, err := mgr.getCheckpointDir(c.ID, options.CheckpointID, true)
+	if err != nil {
+		return err
+	}
+
+	if err := mgr.Client.CreateCheckpoint(ctx, c.ID, dir, options.Exit); err != nil {
```

Comment:
If `options.Exit` is `true`, the container would be stopped by checkpoint process. Do we need to add a `markStoppedAndRelease` here to set `stopped` status of this container and release some resources?

Reply:
Yes, I am not consider that.

Reply:
@xiechengsheng , I re-think the problem, you may not correct for this, `markStoppedAndRelease` will called as soon as container task exit, it is running in a goroutine, called by received containerd exit channel.

Reply:
I'm sorry but could you tell me which goroutine runs this `markStoppedAndRelease` function? :smile:

Reply:
https://github.com/alibaba/pouch/blob/master/ctrd/watch.go#L71